### PR TITLE
Fix issue 18446

### LINKED ISSUE #18446: Add test case for AlterTransactionRuleHandler

### PR DESCRIPTION
**Pull request by tain198127**
Fixes #18446 .

Changes proposed in this pull request:
- add unit tests for AlterTransactionRuleHandler
- fix a potential problems：in shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandler.java 50 lines, must invoke setInstanceContext method before call getResources method , otherwise will get a NullPointerException

Unit test Converage
image

### CODE REVIEW

**Review by terrymanu — CHANGES_REQUESTED** [review]
Could you translate the comment to English?

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandler.java [hunk 1]
```diff
@@ -47,6 +47,8 @@ private void replaceNewRule(final ContextManager contextManager) {
         globalRules.removeIf(each -> each instanceof TransactionRule);
         Map<String, ShardingSphereDatabase> databases = contextManager.getMetaDataContexts().getMetaData().getDatabases();
         TransactionRule transactionRule = new TransactionRule(toBeAlteredRuleConfig, databases);
+        //need add InstanceContext before call getResources
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandler.java:50]: please remove unnecessary comment

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        AlterTransactionRuleStatement ralStatement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
+        when(ralStatement.getDefaultType()).thenReturn("LOCAL");
+
+
+        ConnectionSession connectionSession = mock(ConnectionSession.class, RETURNS_DEEP_STUBS);
+
+        alterTransactionRuleHandler.init(ralStatement, connectionSession);
+        AlterTransactionRuleStatement testAlterStatement = alterTransactionRuleHandler.getSqlStatement();
+        Assert.assertNotNull(testAlterStatement);
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:136]: Please use static import for `Assert`

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+     */
+    @Test
+    public void testGetStatement() {
+
+        AlterTransactionRuleHandler alterTransactionRuleHandler = new AlterTransactionRuleHandler();
+        AlterTransactionRuleStatement ralStatement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
+        when(ralStatement.getDefaultType()).thenReturn("LOCAL");
+
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:131]: Please remove unnecessary  blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+    }
+
+    /**
+     * Method: getStatement()
+     */
+    @Test
+    public void testGetStatement() {
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:125]: Please remove unnecessary  blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        AlterTransactionRuleHandler alterTransactionRuleHandler = new AlterTransactionRuleHandler();
+        AlterTransactionRuleStatement ralStatement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
+        when(ralStatement.getDefaultType()).thenReturn("LOCAL");
+
+
+        ConnectionSession connectionSession = mock(ConnectionSession.class, RETURNS_DEEP_STUBS);
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:133]: Please remove unnecessary  blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        alterTransactionRuleHandler.init(ralStatement, connectionSession);
+        AlterTransactionRuleStatement testAlterStatement = alterTransactionRuleHandler.getSqlStatement();
+        Assert.assertNotNull(testAlterStatement);
+
+    }
+
+    /**
+     * Method: execute()
+     */
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:142]: Test method do not need javadoc

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        ContextManager contextManager = mock(ContextManager.class, RETURNS_DEEP_STUBS);
+        ProxyContext.init(contextManager);
+        AlterTransactionRuleStatement statement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
+        AlterTransactionRuleHandler alterTransactionRuleHandler = new AlterTransactionRuleHandler();
+        AlterTransactionRuleStatement ralStatement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
+        when(ralStatement.getDefaultType()).thenReturn("LOCAL");
+
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:154]: Please remove unnecessary  blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+     */
+    @Test
+    public void testExecute() throws Exception {
+        ContextManager contextManager = mock(ContextManager.class, RETURNS_DEEP_STUBS);
+        ProxyContext.init(contextManager);
+        AlterTransactionRuleStatement statement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
+        AlterTransactionRuleHandler alterTransactionRuleHandler = new AlterTransactionRuleHandler();
+        AlterTransactionRuleStatement ralStatement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:151]: Please remove unnecessary  blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    /**
+     * Method: execute()
+     */
+    @Test
+    public void testExecute() throws Exception {
+        ContextManager contextManager = mock(ContextManager.class, RETURNS_DEEP_STUBS);
+        ProxyContext.init(contextManager);
+        AlterTransactionRuleStatement statement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:148]: Please remove unnecessary  blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        ConnectionSession connectionSession = mock(ConnectionSession.class, RETURNS_DEEP_STUBS);
+
+        alterTransactionRuleHandler.init(ralStatement, connectionSession);
+        ResponseHeader testRespHeader = alterTransactionRuleHandler.execute();
+        Assert.assertNotNull(testRespHeader);
+    }
+
+    /**
+     * Method: getConnectionSession()
+     */
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:164]: Test method do not need javadoc

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    @Test
+    public void testGetConnectionSession() {
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:166]: Please rename `testGetConnectionSession` to `assertGetConnectionSession`

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+        AlterTransactionRuleHandler alterTransactionRuleHandler = new AlterTransactionRuleHandler();
+        AlterTransactionRuleStatement ralStatement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
+        when(ralStatement.getDefaultType()).thenReturn("LOCAL");
+
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:172]: Please remove unnecessary blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    /**
+     * Method: getConnectionSession()
+     */
+    @Test
+    public void testGetConnectionSession() {
+        AlterTransactionRuleHandler alterTransactionRuleHandler = new AlterTransactionRuleHandler();
+        AlterTransactionRuleStatement ralStatement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:169]: Please remove unnecessary blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        when(ralStatement.getDefaultType()).thenReturn("LOCAL");
+
+
+        ConnectionSession connectionSession = mock(ConnectionSession.class, RETURNS_DEEP_STUBS);
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:174]: Please remove unnecessary blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+        alterTransactionRuleHandler.init(ralStatement, connectionSession);
+        ConnectionSession session = alterTransactionRuleHandler.getConnectionSession();
+        Assert.assertNotNull(session);
+    }
+
+
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:180]: Please remove unnecessary blank lines

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+        AlterTransactionRuleStatement ralStatement = mock(AlterTransactionRuleStatement.class, RETURNS_DEEP_STUBS);
+
+        when(ralStatement.getDefaultType()).thenReturn("LOCAL");
+
+
+        ConnectionSession connectionSession = mock(ConnectionSession.class, RETURNS_DEEP_STUBS);
+
+        alterTransactionRuleHandler.init(ralStatement, connectionSession);
+        ConnectionSession session = alterTransactionRuleHandler.getConnectionSession();
+        Assert.assertNotNull(session);
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:177]: Please use static import for `Assert`

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class AlterTransactionRuleHandlerTest {
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java:56]: class should be final if not design for extension

**Review by terrymanu — CHANGES_REQUESTED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandler.java (hunk 1 above)
Comment (tain198127) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandler.java]: fixed

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java (hunk 2 above)
Comment (tain198127) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java]: fixed

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java (hunk 3 above)
Comment (tain198127) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/updatable/AlterTransactionRuleHandlerTest.java]: fixed

**Review by terrymanu — CHANGES_REQUESTED** [review]
The CI still fail, could you have a look?

**Comment by tain198127** [comment]
image

**Pull request merged**